service/worker/workerdeployment/workflow.go

```diff
@@ -116,6 +127,96 @@ func (d *WorkflowRunner) run(ctx workflow.Context) error {
 	return workflow.NewContinueAsNewError(ctx, Workflow, d.WorkerDeploymentWorkflowArgs)
 }
 
+func (d *WorkflowRunner) validateSetWorkerDeploymentRampingVersion(args *deploymentspb.SetWorkerDeploymentRampingVersionArgs) error {
+	if args.Version == d.State.RoutingInfo.RampingVersion && args.Percentage == d.State.RoutingInfo.RampingVersionPercentage {
+		d.logger.Info("version already ramping, no change")
+		return temporal.NewApplicationError("version already ramping, no change", errNoChangeType)
+	}
+	// todo: this will only work when "__unversioned__" is the default current-version of a deployment.
+	if args.Version == d.State.RoutingInfo.CurrentVersion {
+		d.logger.Info("version already current")
+		return temporal.NewApplicationError("version already current", errVersionAlreadyCurrentType)
+	}
+
+	return nil
+}
+
+func (d *WorkflowRunner) handleSetWorkerDeploymentRampingVersion(ctx workflow.Context, args *deploymentspb.SetWorkerDeploymentRampingVersionArgs) (*deploymentspb.SetWorkerDeploymentRampingVersionResponse, error) {
+	// use lock to enforce only one update at a time
+	err := d.lock.Lock(ctx)
+	if err != nil {
+		d.logger.Error("Could not acquire workflow lock")
+		return nil, serviceerror.NewDeadlineExceeded("Could not acquire workflow lock")
+	}
+	d.pendingUpdates++
+	defer func() {
+		d.pendingUpdates--
+		d.lock.Unlock()
+	}()
+
+	prevRampingVersion := d.State.RoutingInfo.RampingVersion
+	prevRampingVersionPercentage := d.State.RoutingInfo.RampingVersionPercentage
+
+	newRampingVersion := args.Version
+	updateTime := timestamppb.New(workflow.Now(ctx))
+
+	// unset currently set ramping version
+	if newRampingVersion == "" {
+		if prevRampingVersion == "" {
+			// no previously set ramping version
+			return &deploymentspb.SetWorkerDeploymentRampingVersionResponse{}, nil
+		}
+
+		rampUpdateArgs := &deploymentspb.SyncVersionStateUpdateArgs{
+			RoutingUpdateTime: updateTime,
+			RampingSinceTime:  nil, // remove ramp
+			RampPercentage:    0,   // remove ramp
+		}
+
+		if _, err := d.syncVersion(ctx, prevRampingVersion, rampUpdateArgs); err != nil {
+			return nil, err
+		}
+	} else {
+		// tell new ramping version it's ramping
+		rampUpdateArgs := &deploymentspb.SyncVersionStateUpdateArgs{
+			RoutingUpdateTime: updateTime,
+			RampingSinceTime:  updateTime,
```

Comment:
RampingSinceTime is a bit weird, because if the ramp % is changing but the ramping version is staying the same, RampingSinceTime will not change, because the time when the version started ramping is still the same.

I would make a variable called `newRampingSinceTime` and do:
```
newRampingSinceTime = updateTime // if this version is newly ramping
if prevRampingVersion == newRampingVersion { // this version was already ramping
  newRampingSinceTime = d.State.RoutingInfo.RampingVersionUpdateTime
}
```